src/script/descriptor.h

```diff
@@ -13,6 +13,17 @@
 
 #include <vector>
 
+// Cache for pre-computed descriptor things
+class DescriptorCache {
+private:
+    std::map<KeyOriginInfo, CExtPubKey> m_xpubs;
```

Comment:
Shouldn't this be a multimap? There can be collisions in the fingerprint + all subsequent path elements.

Reply:
I think that's extremely unlikely. It would require several thousand xpubs in the descriptor itself. I think we have a few other places where we assume origin info are unique.

Reply:
With 93 entries there is a 1 in a million chance for a collision... I agree it's low, but it's not so low that it is negligible, and certainly not so low that one couldn't grind a descriptor to intentionally cause a collision.

This makes me wonder if using KeyOriginInfo as key is the right approach. In theory, it should be sufficient to just key by the index of the key expression in the descriptor, right? As the descriptor is constant, and the cache is specific to one descriptor, that would uniquely identify what the entry is for.

Reply:
>index of the key expression in the descriptor

the what?

Reply:
Every descriptor consists of some functions that combine the output of key expressions (pubkeys, xpubs, origin info, derivation paths, ...). We can assign the successive key expressions numbers 0, 1, 2, ... and use those index positions as key in the map (which would then become a std::map<int, CExtPubKey>).

Reply:
I don't really like using the position to index into the cache. It requires the `PubkeyProviders` to know their position within the descriptor when it really doesn't matter to them.

Reply:
I think that's not a useful argument. If the current way the code works makes it hard to implement the right solution, the code should be changed. It seems to me that using position is the more concise, exact, and efficient way of doing it, and ignoring current code architecture it also seems the simplest, so I think it would be preferable.

You're right that a multimap wouldn't work. Using a KeyOriginInfo to identify the record is inherently lossy - a better solution is just not throwing it away at all.